### ISSUE TITLE
Adding a new view can break previous requests when embedding with column or FK as target

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2277]
As exemplified in the docs: https://postgrest.org/en/stable/api.html#hint-disambiguation

Whenever a view is added it can break previous embeds - which is surprising behavior for most users.

This is specially bad considering we generate a lot of view relationships: https://github.com/PostgREST/postgrest/pull/2262#issuecomment-1107360397. Note that we generate [view relationships recursively](https://postgrest.org/en/stable/api.html#embedding-chains-of-views).

### Proposal

1. When using a column or FK for embedding, only tables should be detected and views not.

```http
GET /orders?select=name,billing_address(name) HTTP/1.1
```

^ Will continue working even when adding more `addresses` views.

(From https://postgrest.org/en/stable/api.html#embedding-disambiguation)

Instead, views should be explicit in the request:

```http
GET /orders?select=name,central_addresses!billing_address(name) HTTP/1.1
```

2. Using a column/FK embed on views will still work for getting tables.

```http
GET /central_addresses?select=name,billing_address(name) HTTP/1.1
```

(the above will get [orders](https://postgrest.org/en/stable/api.html#target-disambiguation))

We had some use cases before doing this https://github.com/PostgREST/postgrest/issues/1230, so it's an organic use case.

---

This will solve https://github.com/PostgREST/postgrest/issues/2238 and https://github.com/PostgREST/postgrest/issues/1643

**Comment by steve-chavez** [comment]
Adding a view of a junction will also make existing many-to-many embeds fail.

For example, adding this to our fixtures:

```sql
create view users_tasks_view as
select * from users_tasks;
```

Results in 12 test failures. For example this one:

```http
GET /users?select=id,tasks(id)&order=id.asc&tasks.order=id.asc
```
So we can take the same approach for junction views - they will only be considered if they're explicit:

```http
GET /users?select=id,tasks!users_tasks_view(id)&order=id.asc&tasks.order=id.asc
```

--- 

The downside is that we'd break this previous use case https://github.com/PostgREST/postgrest/issues/1736


For this we can add a legacy config.

**Comment by wolfgangwalther** [comment]
> When using a column or FK for embedding, only tables should be detected and views not.

This would still allow breaking existing queries when adding a view, namely those queries that use the column name for embedding - and then a view with the name of the column is added later on.

This change would also create more differences between tables and views - and lead to a situation where self-embedding would suddenly not be supported the same on views compared to tables anymore.

Also migrating to schema isolation step-by-step as mentioned in https://github.com/PostgREST/postgrest-docs/issues/531 would be made harder, because some queries might need adjustment at the same time.

> For this we can add a legacy config.

Please don't. Cluttering the code with legacy config options is hardly an improvement. Let's instead make sure that we change it in a way that everyone can migrate to easily.

---

Let's break down on a higher level why adding new views can break previous requests. There's two reasons for that:
- Adding a new view can cause a namespace collision, i.e. by adding a view that has the same name as a column or FK. This would mostly be solved by https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1094611081, except for some collisions between column names and FK names. Once we have a collision, however, we can't resolve it anymore. No disambiguation as currently implemented or designed would fix that. If we wanted to do that, we'd need a way to tell postgrest in the query that a name belongs to a column, a FK or a relation.
- Adding a new view can add a new "path" from origin to target, which creates ambiguity on "how to get there". Those can be resolved by adding hints, but possibly require more than one hint or relationship-type hints (m2m, o2m, m2o). Those conflicts can be made less likely by reducing the number of relationships considered for embedding. The one big thing we can do here is restricting embedding for m2m to strict junctions as mentioned elsewhere. We can restrict it even more as mentioned in https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1094612996.

Note that both situations where an existing query is broken by adding a view can also happen when adding a column or a FK to a table - so those are not really specific to views. Given this and the down-sides regarding schema isolation, I suggest to find a way that does not depend on view vs. table - but rather something like https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1094611081.

**Comment by steve-chavez** [comment]
Hey Wolfgang,

> This would still allow breaking existing queries when adding a view, namely those queries that use the column name for embedding - and then a view with the name of the column is added later on.

That is https://github.com/PostgREST/postgrest/issues/2238 right? My [answer](https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1107664302) here applies then - telling the user to change the view name is fine. It's also a rare case.

> This change would also create more differences between tables and views - and lead to a situation where self-embedding would suddenly not be supported the same on views compared to tables anymore.

Right, self join on views would not be supported. Maybe I can make an exception in that case.

> Also migrating to schema isolation step-by-step as mentioned in https://github.com/PostgREST/postgrest-docs/issues/531 would be made harder, because some queries might need adjustment at the same time.

True. I guess it's a chicken and egg problem. 

1. If views keep working as it is, any new view can break many requests with embedding. Forcing the client to be updated to use a `target!hint` embed.

2. If we make the change, views will need `target!hint`. So migrating a table to a view will also force the client to be updated.

I'd say 1 is much more impactful and surprising for users. Specially considering that we recommend views for more complex queries. So it's a fair trade.

A middle ground would be having a way to mark views for full embedding inference - views that work with col/fk as target.

```sql
ALTER ROLE authenticator SET pgrst.fully_inferred_views = 'a_view, another_view'
```

That would allow users to migrate, we can warn about this setting too. It's one more config, but I don't see another way around it.

**Comment by steve-chavez** [comment]
> Also migrating to schema isolation step-by-step as mentioned in https://github.com/PostgREST/postgrest-docs/issues/531 would be made harder, because some queries might need adjustment at the same time.
> If we make the change, views will need target!hint. So migrating a table to a view will also force the client to be updated.

Actually, the above is not 100% true if the client used the table name as target:

```http
GET /projects?select=*,clients(*)
```

If the `clients` table changes to a view that will still keep working.

If the client used:

```http
GET /projects?select=*,client_id(*)
```

In that case, yes, changing that to a view won't work. The config I proposed would solve that though.

**Comment by wolfgangwalther** [comment]
> Wouldn't that be a much worse breaking change? Also as I mentioned there, users naturally reach for col as target - so we can't break that.

And I think we should break it.

Using columns for embedding might feel natural to some (it doesn't for me - at all) - but really it's just the root cause of all the ambiguities when embedding.

No o2m/m2o embedding that has the relation as target can be broken by adding another view - because the relation is already explicitly specified. Those can only be broken by adding new FKs and that's certainly something that would be expected as you rightly point out.

The same for m2m relationships - as long as those are given as `<relation>!<junction>`, no new view can break them.

Relationships are defined by junction tables and foreign keys - as you can tell from our haskell typing which models relationships this way. And I think we should limit ourselves to just those and throw columns out. This would make embedding so much more consistent overall, much easier to understand, much less error prone.

I'd say:
- Embeddings given as `<relation>` only (without hint) can only embed m2o and o2m relationships.
- Those m2o and o2m can only throw ambiguity errors, when multiple FKs exist between those relations. In this case a single hint can be given with the FK name: `<relation>!<fk>`.
- Strict m2m junctions can be embedded as `<relation>!<junction>`.
- Any other m2m junctions (non-strict) can be embedded as `<relation>!<fk1>,<fk2>`.
- *I still have no idea about a nice way to handle self-embedding*

Yes, this might feel like a "big" breaking change. But really anything we do here will be a breaking change. And once you start thinking in terms of "relations" and "foreign keys" instead of "columns" when embedding, it will not be hard at all. After all an embedding is just a join - and users certainly understand they need a target relation for a join.

One more argument against columns: Using columns does not allow using multi-column FKs as relationships. So it's not a complete feature anyway. PostgREST has made a decision a long time ago, that there is no "simple" route->PK mapping like `/client/:id` etc. - specifically, because those wouldn't support multi-column PKs. Overall, it would be consistent to not use columns here either.

**Comment by steve-chavez** [comment]
> One more argument against columns: Using columns does not allow using multi-column FKs as relationships. So it's not a complete feature anyway. PostgREST has made a decision a long time ago, that there is no "simple" route->PK mapping like /client/:id etc. - specifically, because those wouldn't support multi-column PKs. Overall, it would be consistent to not use columns here either.

I think we're going in circles here. The above was already solved a long time with FK-as-target, which we would also drop with your proposal.

> Yes, this might feel like a "big" breaking change. But really anything we do here will be a breaking change. And once you start thinking in terms of "relations" and "foreign keys" instead of "columns" when embedding, it will not be hard at all. After all an embedding is just a join - and users certainly understand they need a target relation for a join.
> And I think we should break it.
> Using columns for embedding might feel natural to some (it doesn't for me - at all) - but really it's just the root cause of all the ambiguities when embedding.

I've seen many of these cases(e.g. https://github.com/supabase/supabase/discussions/6274), not offering the col-as-target would break those and not offer a path forward - this is what I consider a "big" breaking change.

> This would make embedding so much more consistent overall, much easier to understand, much less error prone.

What I can do to make it more understandable is to deprioritize the col-fk-as-target on the docs and prefer the `target!hint` form to disambiguate. The col-fk-as-target can be left as a last section in the disambiguation part as a "shortcut", possibly even discouraging(maybe deprecating) this form.

I do see a future where col-fk-as-target support is removed, but this change should not be so abrupt.

**Comment by steve-chavez** [comment]
> So we can take the same approach for junction views - they will only be considered if they're explicit:
> The downside is that we'd break this previous use case https://github.com/PostgREST/postgrest/issues/1736
> Relationships are defined by junction tables and foreign keys - as you can tell from our haskell typing which models relationships this way.

Good point. So maybe the above breaking change I proposed is not needed, it's reasonable to fail in that case I believe. Also, the adding of junction views is more of an artificial problem, I have not ever see them reported. The new strict m2m junctions also reduce likelyhood.

> Strict m2m junctions can be embedded as <relation>!<junction>.

I can also prioritize the "relation!junction" form in the docs and leave the old form as a deprioritized shortcut.

> Any other m2m junctions (non-strict) can be embedded as <relation>!<fk1>,<fk2>.

No need to support that now or in the future, flattening should cover those cases https://github.com/PostgREST/postgrest/issues/1233#issuecomment-1115118458. In fact, flattening(as ideated on the issue) would be able to express any m2m relationship - somewhat making the `relation!junction` form unnecessary. Users would still expect a direct way to get m2m relationships though, so I'm not advocating for eliminating that form.

**Comment by steve-chavez** [comment]
> A less disruptive solution to the issue might be just considering the "inferred relationships on views" being 2nd class when in presence of "table relationships". So if we detect 3 relationships and two of them are on views then the table relationship will prevail and be picked. If all 3 relationships are on views then we'd raise the usual error.

From https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1107665475

In light of the opposition to the proposal - the above solution would solve the issue without a breaking change. Though I'd have to see how feasible is to implement it(would probably be less efficient).

**Comment by wolfgangwalther** [comment]
> > A less disruptive solution to the issue might be just considering the "inferred relationships on views" being 2nd class when in presence of "table relationships". So if we detect 3 relationships and two of them are on views then the table relationship will prevail and be picked. If all 3 relationships are on views then we'd raise the usual error.
> 
> From [#2238 (comment)](https://github.com/PostgREST/postgrest/issues/2238#issuecomment-1107665475)
> 
> In light of the opposition to the proposal - the above solution would solve the issue without a breaking change. Though I'd have to see how feasible is to implement it(would probably be less efficient).

Ah, I didn't comment on that, because it seemed like the idea was thrown away already: But I would be even more opposed to this, than to the other suggestion. And that's because with this proposal a working request could not only fail by adding a new view - but it could suddenly **return a different result** (without failing with an error code) when adding a new table relationship that would otherwise conflict with an inferred view relationship, which is already used.

So assume you use an existing inferred view relationship and you're adding a new table relationship that would cause a conflict: Suddenly, because table wins vs view you're now embedding a different relation compared to before, without getting an error.

That's something we should absolutely avoid - and that's why I think we should avoid all "ranking" based approaches like this one. All the different options need to have the same priority.

**Comment by bolerodan** [comment]
Having experienced incorrect embedding very briefly a year or so ago, I'm curious what truly would be a solution to this with the current auto-detection that is in place? I've realized that being as explicit as possible in our queries has been the safest way as possible to avoid this scenario and always using hints. 

Is it even possible to create a solution that guarantees not breaking current query embeds when adding in new tables/views down the line?

**Comment by steve-chavez** [comment]
> I've realized that being as explicit as possible in our queries has been the safest way as possible to avoid this scenario and always using hints.

@bolerodan IIRC you use columns as hints - e.g. `/tbl?select=other!col(*)`- in theory this could break(be ambiguous) if you added a new composite FK from `tbl` to `other` that also included `col`. However the possibility of that happening is low, so it's fine/stable. I believe that's all we can do with the foreign key join(embedding) feature, to minimize the possibility of an ambiguity happening. No perfect solution.

Right now by using the `/tbl?select=col-or-fk(*)` syntax, the possibility of ambiguity is too high in presence of views. This is what I'm aiming to solve with https://github.com/PostgREST/postgrest/pull/2272. In theory there will be breaking changes for pure-view schemas, however in practice I'm not sure if that would cause a lot of breaking changes, since `col-or-fk` is too unstable with views as mentioned before. It's likely that pure-view schemas where already using the more strict `other!hint` syntax.

**Comment by steve-chavez** [comment]
> At the same time we should try to only introduce breaking changes that can be solved by adapting the request you make. We should avoid breaking changes that force you to change your schema - or completely change the way how to query the api in certain scenarios (e.g. forcing you to use "something else" where an embedding worked before).

@wolfgangwalther We haven't been following that criteria for breaking changes. Adapting the request you make involves client updates, which is a much harder change(making sure many clients download the new JS lib version, invalidating cache, etc). In contrast, changing your schema is a much easier change. For example, renaming a constraint(simple change in the db) allowed  a migration path from the old duck-typing bolerodan mentioned. Renaming views is also a simple change(which could be done to avoid a breaking change for col-fk-as-target). Renaming a table/column name would be out of the question of course. Ideally this should be scoped to the `api` schema tailored for postgREST.

**Comment by wolfgangwalther** [comment]
> Adapting the request you make involves client updates, which is a much harder change(making sure many clients download the new JS lib version, invalidating cache, etc). In contrast, changing your schema is a much easier change.

Interesting take. For me, it's quite the opposite.

I can't avoid all breaking changes to my api. To be able to handle **any** breaking changes to my api cleanly, I need to use proper api versioning mechanisms. And that's not just the breaking changes we're talking about here (introduced by PostgREST), but also all breaking changes introduced by my own api development. If I don't have api versioning mechanisms in place, I can never do a proper upgrade.

We suggest to use schema-based api versioning somewhere.

The procedure is as follows:
- Add a new schema with the new api version, including the breaking changes.
- If you want to update PostgREST, run the new PostgREST version and point it at the new schema.
- With a reverse proxy: Redirect your requests to the new version to the new instance, while keeping the requests to the old version at the old instance/schema.
- Update all clients step-by-step to use the new api
- When all clients are updated: shut down the old instance

This is a "rolling upgrade" of a server/client architecture without downtime.

You can automate this in Continuous Deployment setups, too.

Once you have this in place, a change to a client's request is really simple.

> For example, renaming a constraint(simple change in the db) allowed a migration path from the old duck-typing bolerodan mentioned.

Renaming a constraint would be a breaking change to your API, which would force those clients that already use the previous constraint name to update their requests.

> Renaming views is also a simple change(which could be done to avoid a breaking change for col-fk-as-target).

Renaming a view would be a breaking change to your API, which would force those clients that already use the previous view name to update their requests.

> Renaming a table/column name would be out of the question of course.

Renaming a column name (in a relation in the api schema) would be a breaking change to your API, which would force those clients that already use the previous column name to update their requests.

Renaming things does not avoid breaking changes to a client's request **at all**.

**Comment by steve-chavez** [comment]
> Update all clients step-by-step to use the new api

We have always avoided telling users to execute that step.

> Renaming a constraint would be a breaking change to your API, which would force those clients that already use the previous constraint name to update their requests.

Once upon a time, there was duck-typing. Assuming `clients` and `projects` tables, you could do:

```http
GET /projects?select=*,client(*)
```

That is, requesting "clients" without the "s" at the end. That thing caused many more ambiguities that were hard to understand so it was removed. The migration path for that was to rename the autogenerated constraint name(say like `clients_projects_fkey`) to `client` and allow for FK-as-target.

> Renaming things does not avoid breaking changes to a client's request at all.

So yeah, the above is just plain false.

**Comment by steve-chavez** [comment]
> The procedure is as follows:
> This is a "rolling upgrade" of a server/client architecture without downtime.

This is still a good guide that we should document. Managing an additional postgrest instances plus proxy config is definitely not as simple as changing the `api` schema though.

**Comment by wolfgangwalther** [comment]
> That is, requesting "clients" without the "s" at the end. That thing caused many more ambiguities that were hard to understand so it was removed. The migration path for that was to rename the autogenerated constraint name(say like clients_projects_fkey) to client and allow for FK-as-target.

But that was only possible, because FKs were not used at all in the exposed API before. That's not the case here. We're talking about renaming things that **are** exposed already.

> > Renaming things does not avoid breaking changes to a client's request at all.
> 
> So yeah, the above is just plain false.

I wasn't specific enough, I guess. Let me rephrase:

Renaming things that are already part of the API does not avoid breaking changes ... etc.

**Comment by steve-chavez** [comment]
> I think we should strive for a solution that creates as much symmetry as possible, because this is easier to understand well for users and has other benefits (full support for schema isolation would be one).

Full schema isolation support is still there as always, the `select=view(*)` and `select=view!hint(*)` forms haven't changed. In fact the latter one would be a bit better with https://github.com/PostgREST/postgrest/pull/2272 as mentioned in https://github.com/PostgREST/postgrest/issues/2277#issuecomment-1118816064.
 
I'd also argue that disambiguation will be easier to understand with this change, since we no longer have the case mentioned in
https://postgrest.org/en/latest/api.html#hint-disambiguation (new views will no longer cause the disambiguation error). I think it's mostly a matter of rearranging the docs and mention the col-fk-as-target is a convenience.

Another win is that we can now document the self-relationship embedding, without solving https://github.com/PostgREST/postgrest/issues/1643 it was not stable enough to document it, but now it is.

**Comment by steve-chavez** [comment]
Let's see how this goes in the next release. There's still a way to offer backward compatibility with this change, but it would require manually specifying relationships:

```sql
ALTER DATABASE your_db -- could also be ALTER ROLE authenticator 
SET pgrst.your_schema.my_view = $$
{
 "relationships": [
   {"target": "arbitrary_name", "foreign_table": "other", "join_columns": ["my_view.col1", "other.col2"]}
  ]
}
$$
```

Then, it would be possible to:

```http
GET /my_view?select=*,arbitrary_name(*)
```

We should try to keep the "low-code" approach though.

**Comment by steve-chavez** [comment]
Actually implementing the above is not a bad idea for this release. It not only provides a way for users to migrate but also workarounds for other issues like https://github.com/PostgREST/postgrest/issues/1907 

Since it provides a way to do embeds without hints, it can also solve the case where there is a legitimate FK ambiguity, without patching clients.

**Comment by steve-chavez** [comment]
In hindsight, I wonder if the server/db side config for resolving embeds would have been a better choice than the client side hints. Not only it would have kept our syntax simpler, but also allows us to provide better metadata(typings, openapi, etc) since the embeds are static and not resolved at runtime.

@bolerodan Curious on your thoughts on the [above proposal](https://github.com/PostgREST/postgrest/issues/2277#issuecomment-1127914913)

**Comment by bolerodan** [comment]
@steve-chavez could you elaborate a bit more about this proposal?

**Comment by steve-chavez** [comment]
@bolerodan Sorry for the late reply. I gave it more thought and will implement https://github.com/PostgREST/postgrest/issues/2144 to offer a way forward from this breaking change. With #2144 any embedding can be resolved unambiguosly.

**Comment by bolerodan** [comment]
@steve-chavez cool sounds good. I've been keeping my eye on that thread to see how it materializes (hah) on that front.
